### ISSUE TITLE
Support for negative values

### ISSUE THREAD

**Issue by deltaphi** [olliiiver/sml_parser#8]
In #6 I proposed a code change that would let the SML parser correctly handle negative values for two-byte values. In https://github.com/olliiiver/sml_parser/pull/6#issuecomment-1236415248, it was suggested that #7 should actually cover this issue. Unfortunately, quite the opposite is true: the code does now realiably *not* handle negative values for any length of data item. Instead of something like -230 I now get 65xxx - resulting in my code telling me that apparently my solar panel consumes 65kW instead of producing a couple 100 watts.

The change proposed in #6 relied on telling the compiler about the actual data type of the data being parsed and then leaving the conversion handling (in this case: Sign extension, possibly twos' complement converion) to the data type conversion of the compiler. The code from #7 removes all of that information and simply pastes the bit pattern received via SML to a `long long int`.

Switching on the length of the data item being received would be a rather simple approach to implement this. But seeing as this was just removed - what would be the most fitting approach to fix this? If it is only about removing duplicate code, I can try to whip up some template code that should reduce this - although that might not work on every compiler this code passes through.

**Comment by deltaphi** [comment]
See #9 for a proposed change.

**Comment by deltaphi** [comment]
See [Code on Godbolt](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGe1wAyeAyYAHI%2BAEaYxCAAzKQADqgKhE4MHt6%2BekkpjgJBIeEsUTHxdpgOaUIETMQEGT5%2BXLaY9nkM1bUEBWGR0XG2NXUNWc0KQ93BvcX9sQCUtqhexMjsHOaxwcjeWADUJrFuTuPEmKwH2CYaAIIbWzuY%2B4eiCkp1F1e3N14pRrvMbAUCSYq1243QBysN0%2BGywNBCuwAstcABoAfQCAEkhAAVNFCTEALWwuwAbJIYTcIqhPLs8ApQphgExHAA3aYlCBoBjjXa0ATAPkCumCXassSkXbc3leHl4YAhdBShC1MF4ABemDm%2BwA7FDrrtDXSqLsIClNU8ACIHS27DTakx6z5Gl27U4EZYMXZUMRKSHOo2O63Qm4u6UEIV/fl/YIRlgKCLKFkEaJem2mqDRwVZkUEOYaVQADntTzcBzcGfNjwAtLsuNqAFS7QtzOb%2B0NG8OR7PC2O7eMRK1isT7Myk/sJpMEFPEBjtg1G92eieDsBgdMaedBynXVmoPBKzCqFMMdBCeUMCA5nOx8zj8XeTCSruylIKzBK5Aq4hqzUOp0doaeAmhAa70oyzJshy0QQA%2BXhPr%2BWr/vqrqLpgHqzvOLrbiGC6GvwP4QK%2BF4fsqqp4EOVaQnSpaIagVCwWI8FtrEFijlYlh4MhAaoXBjyOmWsS2pmwo3oI%2BaqFQVAOoc5amhRTYtlhgY6sGtyqTuAD0mm7AAYngqikQAatEKQCJ8169qKwLEEo%2BmGegXICDKcrvp%2B358vSBAWF4UnRCYACsFjIuiWK4viRKXAFlrcYBuzEW5ZE/hRG7KYaCWKklhpVkOECxnMtBeT5fnEIFFh4IF1qsTx3a1X2D5Dpu1VxQRpoZaRX6qgAno11E9XJVGsbsXXWNYsV4a62lSgI7J1MOj41S6U1CIiAQYgA8gA4hA5hmDl6bmAFeCSjkQ6HcduwRIQZ1mEdgVuHOZhmJKVaSv1lh1qQi1GlNqGmoWuxNmaGo1nWra7LWEAA02XWtmlroNQJ6ZXqJVl5oV4zFVQ/lBSlH3vWxXCVd9qHlnJUDQ6aOW1vW2qQ5TsMsShKlqTV2ldpZMaigOU4zmmQmmgWxaluTwOatW9aA82cPNRN7PORGnOCn2A5DojY4rrzqZbnF2nAaaqtrhuyG7FNEEsng7LzfBbOafreUMkyFvskUnIPi9IPg46bFm07UHW5gtsukrubeoVCRa7OvWy0tOl/a1eWito0dscnA0g9R2ijZY42x39rpUOHkdpjqgm2gWum6SLhymsnilM0Hf04XLceGhzqNcxGb4MNgx6uMXQ4AH6F3gEfJtrsuN7xI5IwL3e9ye6DFzrE359p6tWALRNDVN1zoEq9FUEoEatQQADuqBgBwCjTSwCT0Gwgi283LpHov54KoxtAe3%2BK8ukuUcHxbg0rhKabhlinFFCZWyaQLId2VtZWoSgnI8gjO1dyqoMbeV8tjUqQUQoYmxHiAkxJKq5yNOgrKKUBZNWZulVymVOo/kQrlfKWCsY43KsTGORoQ71RnjQ%2BGCdKFMOGinMR6dNR9WzhYchqEprclmhGPiJNTY6RWmtAIW0dpPX2gLc6J1kg3SOpKK6EYDq3QqgFB6u0f4IQJp9VRv1UJQylmLUGtMIb/SlozeGLpEal2Rnw8S7CcGcLxmxBxRNoqqOwrJGuFM3HUzBnTbxMMZZ0N1KzOKACvRAMniAr41wpq91YPfR4rUzDVgiF1FMAc4FRjRrsGySgzAoJct3DqHlQklTKgQsKxDIpkN1Jkl8DCumqj0baHpuCyqbhiTww0TBXjRAIO4q06YzANzirGLgpI0TKIEcJXZ%2Bz0ZFTCXgiwZhKq6grInAgeyDkFXOb0oK0TbRkxroWeGuT5rAOydcER3SXm4OicFVEhDwokKih8/Reo7SqA0M9BF2NJQFioIObcO8dLVjHETb44yME/hmdEa5%2BCIUDIiqQ6KZ14UFiRWixFzQEVmGQFk6iU1YgaE3LhFWTBggQDkWCAg6AQAoCWOY%2BJFZdrrWIPKYII4xDABIIQBALAQCjjMNXCsLTMBXhBdESWnzpVPT0gZUiSqVUEDVRq3a2rmlIMwPZD8%2BrMYXKNVKzVuwPDEFOA4XYRkmKPFtaa41DrbKYDaSS4gHrBIVnBGK1wtB4YJvFV4SVcbNWyvlcwWguxLVyuteqr1YbdWuuwSVLZ9q7XOqVAW1Vxa7WlsdbW8tHDiBVrDXan1fqIyBsfCGrVzaI1RoNR2mSmbU1JpTSKsVaB03VqevpBg9IEAfgAHRNs9VO08yaCmWg4AsWgnAAq8D8BwLQpBUCcDLJYawYIlgrH4mYWIPBSAEE0IehYABrOIsR12xEA0B4DwHST6E4JIM9n6r2cF4AoEAGh32foWHAWASA0B3zoNEcglAMMJCwzEYAXAnp8DoDOeDEAIjQauswYgXVOBvow4/Ag60GC0Hoxe3gWAWCGGAOITjpB8C9stpgeDAmjwVHTWsS9sZWjQcKhEYgtQuoeCwAx3gBA5Xqu4IevgBhgAKCMngTAZ91oJEYOpmQggRBiHYFIKz8glBqGg7oZoBgjAoBkfoPAER4OQAWKgBI7QxNwdaBUdoLhTwjCaKQQIUxXb9GaDkVIAhovZGSClhgPQEsxDGGFyoAhOjDE8I0PQ5QCsdAmNlvouXBhdDS2MKr8WasSAWAoR9qw9CacwGsHgR6T1QYE9ejgRZSTVnJLsYAyA2XEfXVqiAt6OI2F2LgQgJBRyvslB4TD9AfwbHrBp5DP6/0AZA2dwDYHj0cEg6Qc9l7htwYQ0hzjKGYCIBABKhI6acMQA/GtuVormj8Gs6IcQ9ngeOZUOoATrnSBnyUwkdT/WOCntu9B4b6101fYjPRXYo3xuSEm9NusZg5umm2/h3bG2DvPa0K2Ugv7AOnfOyB8D13Bv3dg7YJ7H6XvI7MBz3gD3adftILNMyfhJBAA) if you want to toy around with it.

**Comment by olliiiver** [comment]
Hi @deltaphi.

Thanks for coming back and spending time on this.

Let me understand how this works...

I haven't been aware that negative values are also a thing with energy meters, but its mentioned in the documentation:

https://www.bsi.bund.de/SharedDocs/Downloads/DE/BSI/Publikationen/TechnischeRichtlinien/TR03109/TR-03109-1_Anlage_Feinspezifikation_Drahtgebundene_LMN-Schnittstelle_Teilb.pdf?__blob=publicationFile&v=1 -> 6.2

So if the last bit is 1, it's a negative number and all bits have to be inverted?

0xfefb would become 260 or 261? 

```
int16_t a = 0xfefb;
a =~ a; // bitwise NOT 
std::cout << a << std::endl; // 260
```

Oliver

**Comment by olliiiver** [comment]
I think we must also check the Type-Length-Field -> signed vs unsigned value. 

Can you provide a dump output from the OBIS value?

**Comment by deltaphi** [comment]
Hi @olliiiver,

this is bog-standard how to represent negative integer values. In a signed value, the highest order bit is the sign bit. If the highest order bit it 0, it's a positive value. If it is 1, it's a negative value. There is then multiple ways how the remainder of the value has to be interpreted: Is it an absoulte value and you just put the sign in front of it? In this case, `0xfefb` would be -32507. Or do you invert all the other bits for a negative value? This is more convenient for computation - for details, I recommend reading up on [Integer Representation on Wikipedia](https://en.wikipedia.org/wiki/Integer_(computer_science)) - the german Wikipedia Article on this might actually be easier to grasp.

From context, I can tell that my smart meter uses a complement representation. There is no way I am either providing or drawing 32kW - so it must be -260 or -261. I don't actually know whether my meter uses ones' complement or twos' complement, and at the current point in time it does not actually matter. I'm fine with the possible off-by-one and it does not actually affect the code in this PR. All the code does is to perform sign extension.

To elaborate on this a bit: say we use twos' complmenent, so 0xfefb (16bit value) represents -261. The bit 0x8000 is set, all the other bits have to be inverted (and added a decimal 1) to find the correct absolute value (261). If we want to cast the 16bit value to a 32bit value, what is the correct bit pattern to use? Intuitively, you might consider 0x0000fefb. However, as we look to the highest bit for the sign, the sign bit has now changed to 0, so this would be a positive number - can't be the solution. Another intuitive solution is to move the sign bit. This would give us 0x80007efb. However, if we apply the conversion rules, this gives us an absolute value of 0x7FFF8105 - and 2.147.451.141 Watts is even more off than before  (at this point, maybe the wires in the walls start to glow or outright evaporate?). So the correct bit pattern is 0xfffffefb - if the sign is negative, you fill up with ones (and if it is positive, fill up with zeros).

**Comment by deltaphi** [comment]
> I think we must also check the Type-Length-Field -> signed vs unsigned value.
> 
> Can you provide a dump output from the OBIS value?

I can look into providing a debug output from SML Parser tomorrow - is that what you are looking for?

However, I would really appreciate an override parameter. Smart Meters seem to have their own mind when it comes to what data they are sending. For example, the multiplier value for the total power on my meter is consistently off by three for my meter - i.e., in an SML message it claims to report kWh when it is actually reporting a Wh value.

**Comment by olliiiver** [comment]
Great. Yes, some hex dump (the serial debug output the lib provides) would be fine.

Is your type of meter available on https://github.com/devZer0/libsml-testing ? Would also like to implement a test for negative numbers ...

**Comment by deltaphi** [comment]
Hello @olliiiver ,

the Meter is an ISKRA MT631. You can find an example dump in [this Gist](https://gist.github.com/deltaphi/d2f13c6dc1c02cff488820a403fe470b). In this dump, the power value, 16.1.0, is -12.

**Comment by deltaphi** [comment]
Fixed by #9
